### DISCUSSION TITLE
PorousFlow: output the mass_frac of one component in PorousFlowMassFraction material as real value

### DISCUSSION THREAD

**Discussion by Joseph-0123** [idaholab/moose#23526] [Q&A Modules: General] [answered]

**Comment by cpgr** [comment, marked as answer]
So the mass fractions should be nonlinear variables in your example so should be in the exodus file (unless you mean something else?)

**Reply by cpgr** [reply]
In any case, you don't need to modify the code to output any of these things. `PorousFlowPropertyAux` gives access to virtually every property so you can easily save these to AuxVariables and then do whatever you like with them.

**Reply by cpgr** [reply]
OK, that's not what I thought you meant!

In your case, you don't need to change `PorousFlowMassFraction` at all. It provides a material property that is a vector of mass fractions of all components in all phases. You can simply get that material property in your kernel and then pick out that phase and component that you want. The `PorousFlowMassTimeDerivative` kernel does exactly what you are trying to do so take a look at that code for inspiration.

**Reply by cpgr** [reply]
Something that I forgot is that the properties are evaluated at the nodes in that `PorousFlowMassTimeDerivative` but you most likely want them at the quadrature points in your kernel. So replace `_nodal` with `_qp` in the property name and `_i` with `_qp' when using the property!

**Reply by Joseph-0123** [reply]
> Something that I forgot is that the properties are evaluated at the nodes in that `PorousFlowMassTimeDerivative` but you most likely want them at the quadrature points in your kernel. So replace `_nodal` with `_qp` in the property name and `_i` with `_qp' when using the property!

Ok, I will try that. Thanks again.

Have a nice weekend,
J

**Reply by Oops-Qiao** [reply]
ok

**Reply by cpgr** [reply]
To allow an arbitrary number of components in an arbitrary number of phases, we construct a 2D vector of mass fractions in `PorousFlowMassFraction` - basically `[[ comp 0 in phase 0], [comp 1 in phase 0]], [[0 in 1], [1 in 1]]` for a two-component, two-phase system. These mass fractions could be non-linear variables if they change during the simulation, or could be constant aux variables. This material property is then available in any other kernel, material, AuxKernel etc through the normal material property interface.

So, `PorousFlowMassFraction` takes the mass fractions provided in the `mass_fraction_vars` parameter, makes a material property `PorousFlow_mass_frac_qp`, which a kernel can access using 
```
_mass_fractions(
        getMaterialProperty<std::vector<std::vector<Real>>>("PorousFlow_mass_frac_qp"))
```
and the kernel can use the values (that were provided to `PorousFlowMAssFraction` in the `mass_fraction_vars` parameter) by indexing into the 2D vector like so for phase and components
```
_mass_fractions[_qp][phase][comp]
```

We do it this way so that we only have to list the mass fractions in one material in the input file, rather than having to fill them out in multiple places (materials, kernels, auxkernels, post processors).

As for your second question about no `_u` in the kernels, we made a choice to make the kernels so that we could reuse them no matter what the type of the nonlinear variable was. There are several possible choices for the independent variables for multiphase, multicomponent flow. Consider the 2 component, 2 phase case. You can have 2 pressure variables, a pressure and saturation, a pressure and capillary pressure, etc. By not using the variable values directly in the kernels,  the same kernels can be used for any variable choice. We use the materials to pass the variable information through to the kernels.

Hopefully that makes sense?

**Reply by Oops-Qiao** [reply]
Wow, Thanks a lot!!!


Regards,
Jerry

**Reply by cpgr** [reply]
Pretty much. If we didn't do that we would need a version of each kernel for a pressure variable, a version for a saturation variable, a version for a mass fraction variable etc.

**Reply by Oops-Qiao** [reply]
Thanks very much Chris!

**Reply by cpgr** [reply]
To get that I think you would need to alter the code to write it out to the console. I don't think there is an easier way sorry.

**Reply by Oops-Qiao** [reply]
> Pretty much. If we didn't do that we would need a version of each kernel for a pressure variable, a version for a saturation variable, a version for a mass fraction variable etc.

[ICs]
  [./hydrostatic_ic]
   type = FunctionIC
   variable = porepressure
    function ='1e6-1e6*x'
  [../]
[]

**Reply by cpgr** [reply]
Perhaps a `PorousFlowOutFlowBC` that lets the fluid with component escape would be a better choice? See https://mooseframework.inl.gov/modules/porous_flow/boundaries.html

I'd be happy to take a better look if you can post a basic input file.

**Reply by Oops-Qiao** [reply]
> Perhaps a `PorousFlowOutFlowBC` that lets the fluid with component escape would be a better choice? See https://mooseframework.inl.gov/modules/porous_flow/boundaries.html
> 
> I'd be happy to take a better look if you can post a basic input file.

Thanks @cpgr , it works!

**Reply by cpgr** [reply]
This plot is basically zero C2 being below the tolerance of the simulation. If you tighten the tolerance criteria it would probably get even smaller. 

Did you add kernels for the new component C2? And add it to the mass fraction material?

**Reply by Oops-Qiao** [reply]
> This plot is basically zero C2 being below the tolerance of the simulation. If you tighten the tolerance criteria it would probably get even smaller.
> 
> Did you add kernels for the new component C2? And add it to the mass fraction material?

Thanks Chris, I see. well, I did not use new kernels and material for C2. Perhaps using new kernels for C2 could fix it, right?

**Reply by cpgr** [reply]
For a multicomponent model, each fluid component needs their own kernels for mass conservation and transport. Otherwise, they aren't solved for at all.

**Reply by cpgr** [reply]
It's because you never fill `_mf_vars` in the constructor as you commented it out. So it is never resized, so doesn't have any elements. When you try and access it you get a seg fault. 

I still think that this is uneccesary. You are going to get your material `newcom` in a kernel or other object - just get the mass fractions through the material property interface do this logical `if () else ()` there?

**Comment by cpgr** [comment]
Maybe we should step back and look at what you are trying to achieve? If you could describe it a bit perhaps we can come up with a better solution?

**Comment by cpgr** [comment]
Also - I'm getting super confused by the two discussions here - can we spin them out to two threads?

**Reply by Oops-Qiao** [reply]
Ok, thanks Chris.

**Reply by cpgr** [reply]
You are probably violating mass conservation with that change. Like Joseph, perhaps some details of what you are trying to achieve so we can come up with a better solution.

**Reply by Oops-Qiao** [reply]
> You are probably violating mass conservation with that change. Like Joseph, perhaps some details of what you are trying to achieve so we can come up with a better solution.

OK， thanks for your quick reply Chris.
it is fixed now
